RecoEgamma/EgammaTools/src/HGCalEgammaIDHelper.cc

```diff
@@ -25,11 +27,13 @@ void HGCalEgammaIDHelper::eventInit(const edm::Event& iEvent, const edm::EventSe
   iEvent.getByToken(recHitsFH_, recHitHandleFH);
   edm::Handle<HGCRecHitCollection> recHitHandleBH;
   iEvent.getByToken(recHitsBH_, recHitHandleBH);
+  edm::Handle<std::unordered_map<DetId, const HGCRecHit*>> hitMapHandle;
+  iEvent.getByToken(hitMap_, hitMapHandle);
 
   recHitTools_.getEventSetup(iSetup);
   pcaHelper_.setRecHitTools(&recHitTools_);
   isoHelper_.setRecHitTools(&recHitTools_);
-  pcaHelper_.fillHitMap(*recHitHandleEE, *recHitHandleFH, *recHitHandleBH);
+  pcaHelper_.setHitMap(const_cast<std::unordered_map<DetId, const HGCRecHit*>*>(hitMapHandle.product()));
```

Comment:
Casting away `const` of  input event products is not allowed.

Reply:
@makortel do you mean that I should move the `const_cast` like [here](https://github.com/cms-sw/cmssw/commit/cfcd7da245570e0c73fae960574848f183340fd5) or that I should never use it?
If the latter, [this](https://github.com/cms-sw/cmssw/pull/30061/files/f5f59f96fe6b23499fc120293838ed9e6f99f77f..cfcd7da245570e0c73fae960574848f183340fd5#diff-aab1c6f6feeda84bbf94d90c54b3a408R95) `hitMap_` can not be `const` because [this method](https://github.com/cms-sw/cmssw/pull/30061/files/f5f59f96fe6b23499fc120293838ed9e6f99f77f..cfcd7da245570e0c73fae960574848f183340fd5?file-filters%5B%5D=.cc#diff-087a9dbe46a3a27810ba50a90ef2dc28R43) modifies it. How can I work it out?

Reply:
If it modifies it, you must make a copy. Under no circumstances are any modules allowed to change data in the Event, LuminosityBlock, Run or EventSetup.

Reply:
@Dr15Jones is [this](https://github.com/cms-sw/cmssw/commit/cfcd7da245570e0c73fae960574848f183340fd5#diff-087a9dbe46a3a27810ba50a90ef2dc28R34) copy OK or shall I define a `const *hitMap_`, assign it the event product and then copy the `hitMap_` to a non-const instance?

Reply:
So the problem appears to be `fillHitMap` since that appears to be the only function in `EGammaPCAHelper` that actually modifies the content of the map. (in addition `setHitMap` leaks memory). So is `fillHitMap` needed by this code?

Reply:
It looks like you removed the use of `fillHitMap` with the call to `setHitMap`. In addition, only `HGCalEgammaIDHelper` uses `EGammaPCAHelper` which means now no piece of code calls `fillHitMap`. Therefore you can safely remove that function and change the internal variable from
```C++
std::unordered_map<DetId, const HGCRecHit *> *hitMap_;
```
to
```C++
std::unordered_map<DetId, const HGCRecHit *> const* hitMap_;
```